Editor Menubar Example: Improve high contrast support and refactor Javascript

jongund opened the pull request:
closes #1355

This pull request is to update the editor menubar example.

[Editor Menubar Preview link](https://raw.githack.com/w3c/aria-practices/issue1355-update-menubar-editor-js/examples/menubar/menubar-editor.html)

The goal of the update is to:
1. Simplify the Javascript code and use a single object
1. Improve high contrast support
1. The Javascript APIs will make it easy to use with ARIA AT setup scripts

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65
* [ ] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @a11ydoer
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @charmarkk 
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @zcorpan

jongund commented:
@spectranaut 

There is a problem with the regression test "key.SPACE" feature, it does not seem to be generating the correct code for testing.   I had to add some code on line 524 to pass the regression test for the 
 "space" character test.   Can you look into this issue?

smhigley commented:
@jongund I pushed a commit that fixes the failing spacebar test, based on my comment in #1358. All the tests pass locally for me now :)

zcorpan commented on examples/menubar/js/menubar-editor.js:
```diff
@@ -0,0 +1,810 @@
+      role = node.getAttribute('role');
+
+      if (role) {
+        role = role.trim().toLowerCase();
```
I think both `trim()` and `toLowerCase()` are technically incorrect to normalize the `role` attribute, but the spec isn't very clear: https://github.com/w3c/aria/issues/1236

I think this entire `if` block can be removed.

jongund commented:
Questions:
1. Should there be a 1px border on the menubar when focus is not in the menubar
1. Convert the icons to SVG for more control over color when a menu item receives focus and in high contrast modes

jongund commented:
@charmarkk 

I added a 1px border to the menubar. could you redo your review and comment on the high contrast mode with the boarder and let me know if you think it is better than no border when the none of the menuitems in the menubar has focus.   When a menuitem has focus in the menubar the border will get thicker (3 pixels).

a11ydoer commented:
I tested with Safari(13.1), Chrome(version 80), and Firefox(75.0) on Mac(10.14.6 Mojave). 

@jongund 
One problem I had with keyboard navigation in Safari(version 13.1) is that my focus directly goes into text area, not menu bar. Once I move my focus into menubar with mouse, keyboard navigation works as documented.

jongund commented:
@mcking65 
@carmacleod 

Carolyn, can you take a look at the changes I made to the menubar editor example:
1. Restored mouse hover behavior to current example
1. Updated styling to look more like a web editor (additional feedback welcome)
1. Converted images for checkbox and radio to SVG format

carmacleod commented on examples/menubar/menubar-editor.html:
```diff
@@ -47,82 +43,82 @@ <h1>Editor Menubar Example</h1>
+      <div class="menubar-editor">
+        <ul role="menubar" aria-label="Text Formatting">
+          <li role="none">
+            <span role="menuitem" aria-haspopup="true" aria-expanded="false" href="#">Font</span>
```
```suggestion
            <span role="menuitem" aria-haspopup="true" aria-expanded="false">Font</span>
```

carmacleod commented on examples/menubar/menubar-editor.html:
```diff
@@ -47,82 +43,82 @@ <h1>Editor Menubar Example</h1>
+            </ul>
+          </li>
+          <li role="none">
+            <span role="menuitem" aria-haspopup="true" aria-expanded="false" href="#">Style/Color</span>
```
```suggestion
            <span role="menuitem" aria-haspopup="true" aria-expanded="false">Style/Color</span>
```

carmacleod commented on examples/menubar/menubar-editor.html:
```diff
@@ -47,82 +43,82 @@ <h1>Editor Menubar Example</h1>
-          </ul>
-        </li>
+          <li role="none">
+            <span role="menuitem" aria-haspopup="true" aria-expanded="false" href="#">Text Align</span>
```
```suggestion
            <span role="menuitem" aria-haspopup="true" aria-expanded="false">Text Align</span>
```

carmacleod commented on examples/menubar/menubar-editor.html:
```diff
@@ -47,82 +43,82 @@ <h1>Editor Menubar Example</h1>
-        </li>
-      </ul>
+          <li role="none">
+            <span role="menuitem" aria-haspopup="true" aria-expanded="false" href="#">Size</span>
```
```suggestion
            <span role="menuitem" aria-haspopup="true" aria-expanded="false">Size</span>
```

jongund commented:
@mcking65 
@zcorpan 

When I run the regression test locally I don't get any errors, but the CI is finding to failed tests.
I have updated the npm installation and rerun the tests locally and everything passes, but it makes no difference they still fail in Github.   
So what should I do?

Tests failing in CI:
1. "menuitemcheckbox-aria-checked"
2. Something with the keyboard, maybe "submenu-character"

jongund commented:
@carmacleod 

Carolyn, I have fixed the validation bugs realted to `href`, but there is a problem somewhere with the regression tests, they pass for me on my local machine, but the CI is finding two failures, frustrating.

jongund commented:
@mcking65 
@zcorpan 

Looks like the extra awaits fixed the issue with the regression test not working in the CI process for the menubar editor updates.

shirsha commented:
- In the updated menubar, the menu's that have sub-menus have no visual indicator as such.

- The menu closes whenever user selects. They must open it each time to make more selections. Ex: Style/color. The user can select more than one style from the available options but the  menuitem closes as soon as user selects any one option. Similarly, when user select "Smaller" in Size by default "medium” font size is selected. To change the font size, the user must open the Size menuitem and change the selection.

- The red color text (16px) against white background doesn't pass the required color contrast ratio of 4.5:1

jongund commented:
@mcking65 
I fixed the merge conflicts, but the regression test file `test\tests\menubar_menubar-2.js` still needs to be removed, the example it once tested has been changed to `menubar_menubar-editor,js`.  I tired to remove the file, but it generates a merge conflict, so I put it back in.  Maybe you can fix it.

jongund commented:
@mcking65 
As I thought the regression test `test\tests\menubar_menubar-2.js` is failing, because the example it references does not exist anymore (e.g. new example name `menubar-editor.html` and new test file `test\tests\menubar_menubar-editor.js`).  When I delete the uneeded test file from the branch it gives the merge conflict in GitHub you reported earlier, so I am not sure what to do fix the problem.  Seems like a catch 22 for me.

mcking65 commented:
@jongund, what if we keep the current editor menubar and its tests; just make this PR about adding new versions with new names/location. Then, after we merge this, we do a separate PR to remove the unnecessary files? I wonder if that would solve the problem.

jongund commented:
@mcking65 
Restored original menubar editor example files, seems like that should fix the merge conflict problem.

jongund commented:
@mcking65 
Looks like it is ready to merge now.
